### PR TITLE
using pinched grid for numerical aquifers.

### PR DESCRIPTION
**Pull request by GitPaean**

### CODE REVIEW

**Comment by GitPaean** [comment]
serial results look fine while parallel results are wrong.

File: opm/grid/cpgrid/processEclipseFormat.cpp
Comment (GitPaean) [inline opm/grid/cpgrid/processEclipseFormat.cpp:282]: Only process zero can reach here. It means other processes might not have this updated NNC information.

Comment (GitPaean) [reply opm/grid/cpgrid/processEclipseFormat.cpp:282]: It is fixed in the ebos side.

**Comment by GitPaean** [comment]
the parallel results looks fine now. fixing all the tests now.

**Comment by GitPaean** [comment]
In this PR, we process the numerical aquifers connection information in the middle of the grid processing, which does not look natural, but I do not think it is an easy task to refactor to do the PINCH, MINPV in the parser.  I am marking this PR to be ready for review.

File: opm/grid/cpgrid/CpGridData.hpp
Comment (blattms) [inline opm/grid/cpgrid/CpGridData.hpp:215]: Your really need to document this properly, as the change seems a bit dangerous. We seem to call this function with an empty ecl_state, which is fine as long as there are no numerical aquifers.

The danger is that in the future fellow developer might assume that the ecl_state is always valid and use it for other things. In that case they will have a hard time debugging.

Comment (joakim-hove) [reply opm/grid/cpgrid/CpGridData.hpp:215]: This looks very dangerous to me - the empty `EclipseState` is purely an artifact of the serialization - it is deeply invalid and should not be used for anything!

Comment (GitPaean) [reply opm/grid/cpgrid/CpGridData.hpp:215]: I will add the document here. If we do not use `EclipseState` here, we will need the `EclipseGrid`, `FieldPropsManager`, `AquiferConfig` and `inputNNC` of the `EclipseState`. I really do not want to add so many arguments and passing them around. One of the big difficulties when developing this part is that we have five different functions with the same name `processEclipseFormat`, with different arguments, and doing different but related things, and call one other.  Maybe it is good idea to rename the functions to get a little bit more descriptive (maybe I should have done it before developing this PR). Anyway, it is not related to this PR directly. 

> We seem to call this function with an empty ecl_state,

What do you mean here?

Comment (GitPaean) [reply opm/grid/cpgrid/CpGridData.hpp:215]: The empty `ecl_state` happens in the following code in CpGrid.cpp. 
```c++
    void CpGrid::processEclipseFormat(const grdecl& input_data, double z_tolerance,
                                      bool remove_ij_boundary, bool turn_normals)
    {
        // TODO: just fix the interface, not sure it will be problematic
        Opm::EclipseState ecl_state;
        Opm::Deck deck;
        using NNCMap = std::set<std::pair<int, int>>;
        using NNCMaps = std::array<NNCMap, 2>;
        NNCMaps nnc;
        current_view_data_->processEclipseFormat(input_data, ecl_state, deck, nnc, z_tolerance, remove_ij_boundary, turn_normals);
```
Do you think to use pointers here better than using the references? Then we can just pass in `nullptr` for the above situation.  Please comment.

Comment (joakim-hove) [reply opm/grid/cpgrid/CpGridData.hpp:215]: Yes I think using a pointer or optional is better here - as I said - things will go fast to pieces if you use a default constructed EclipseState

Comment (GitPaean) [reply opm/grid/cpgrid/CpGridData.hpp:215]: Okay. I will use pointers here. I think optional to reference is not there yet, right? Correct me if I am wrong here.

Comment (blattms) [reply opm/grid/cpgrid/CpGridData.hpp:215]: Wouldn't std::optional make a copy of EclipseState? That might be expensive...
You would need to use a std::optional with a std::reference_wrapper<EclipseState> inside to prevent the copy. Not sure whether that really works.

Comment (joakim-hove) [reply opm/grid/cpgrid/CpGridData.hpp:215]: >You would need to use a std::optional with a std::reference_wrapper inside to prevent the copy. Not sure whether that really works.

That works - I have tried it in another context; but maybe slightly over engineered? I would go with a pointer, but this is not my turf.

File: opm/grid/cpgrid/CpGrid.cpp
Comment (GitPaean) [inline opm/grid/cpgrid/CpGrid.cpp:499]: not sure why `nnc` can not use `{}` as the original code, will try more.

File: opm/grid/CpGrid.hpp
Comment (GitPaean) [inline opm/grid/CpGrid.hpp:266]: One thing unfortunate due to the change here is that, originally the main input is `EclipseGrid`, but now the main input `EclipseGrid` becomes something hidden in the `EclipseState`. 

Maybe we still can keep `EclipseGrid` as the main input, but also has `EclipseState` there for extra information. But we can not guarantee the `EclipseGrid` is the grid from `EclipseState`.

File: opm/grid/cpgrid/processEclipseFormat.cpp
Comment (GitPaean) [inline opm/grid/cpgrid/processEclipseFormat.cpp:287]: @joakim-hove , it is here the `mutableNumericalAquifers()` will be modified.

**Comment by GitPaean** [comment]
rebasing turned out to be a little tedious. So merging is used here to fix the conflicts.

File: opm/grid/cpgrid/CpGridData.hpp
Comment (GitPaean) [reply opm/grid/cpgrid/CpGridData.hpp]: This change really backfired in the downstream modules. in opm-upscaling module, it only looks at the eclipse_grid. With the same grid deck, you can generate a EclipseGrid and also possible for a EclipseState with an EclipseGrid. But these two grid are different, because since there is no pore volumes, all the grid cells will be marked in active.

File: opm/grid/CpGrid.hpp
Comment (GitPaean) [reply opm/grid/CpGrid.hpp]: I will recovered to be the main argument is the `EclipseGrid`. `EclipseState` will be secondary argument to provide other information. When there is only `EclipseState` provided, we use the `EclipseGrid` belonging to `EclipseState`.

**Comment by GitPaean** [comment]
This is the major parts related to this effort to get this piece of numerical aquifers in.  Besides some code related to numerical aquifers directly, the main part is interface changes related to functions `processEclipseFormat`.  Please review and suggest. 

The jenkins was triggered from upsctream PR OPM/opm-common#2389 and is green now.

**Review by bska — COMMENTED** [review]
Looks good to me.  I just had a couple of minor questions.  Once those are addressed, I'm prepared to merge this along with its upstream OPM/opm-common#2389.

File: opm/grid/cpgrid/processEclipseFormat.cpp
Comment (bska) [inline opm/grid/cpgrid/processEclipseFormat.cpp:254]: I didn't notice this before.  Strictly speaking, this particular function is not protected by `HAVE_ECL_INPUT` so it was originally supposed to be callable without types from opm-common.  I'd be happy to remove that restriction, but maybe we should discuss this topic a little more before we merge the PR.

File: opm/grid/cpgrid/CpGrid.cpp
Comment (GitPaean) [reply opm/grid/cpgrid/CpGrid.cpp:548]: Not totally sure that I understand your question correctly here. 

For this situation, nnc will always be empty. The only reason it can be modified within this function call is numerical aquifers, but we pass in `nullptr` for `EclipseState*`, so no numerical aquifers.  `nnc` created here is purely for interface.  It does not look pretty.  

Suggestion is very welcomed here.

File: opm/grid/cpgrid/processEclipseFormat.cpp
Comment (GitPaean) [reply opm/grid/cpgrid/processEclipseFormat.cpp:254]: Sure. I am not that familiar with grid related stuff. 

In my own opinion, the grid processing (getting the final ACTNUM, we should not drop any information before determining the final ACTNUM, so a cell is free to change from inactive to active.) should be done in the early stage of the parsing, so all the numerical aquifers stuff can stay in opm-common.  

The current way is intruding because we do the processing (can remove more cells) after parsing, while parsing needs processed grid and parsing information also intervene the grid processing.

**Comment by blattms** [comment]
Thanks for adding documentation. My main point for requesting it, was that developers can see which are optional or which do only have to have a limited state in certain circumstances. Seems like that is still not clear from the text. Or maybe this did change?

I would like to get the following info from the documentation:
1. If and when I am allowed to pass nullpointers
2. If and when I allowed to pass incomplete objects (e.g. EclState not reflecting the whole model)

Or has that changed meanwhile.
Would be cool if that was be easy to find. Otherwise I concur with @bska's opinion (not even sure whether I have the required knowledge to do so).

**Comment by blattms** [comment]
I am approving now as this could be done in a separate PR (which often means it will slip).

**Review by blattms — APPROVED** [review]
Apart from doc improvements this looks fine.

**Comment by GitPaean** [comment]
> If and when I am allowed to pass nullpointers

Yes. Null pointers are allowed, which makes it okay to only have the grid passed in.  When you do not have `ElcipseState`  or ` `Deck` available, you can pass in null pointers. This is referring to the function `void CpGridData::processEclipseFormat` in `processEclipseFormat.cpp`. 

> If and when I allowed to pass incomplete objects (e.g. EclState not reflecting the whole model)

This is hard to tell. For example, if a deck only have the GRID description without providing `PORO` or `PORV` information, all the grid cells will be marked inactive. Then it will probably not work very well.  But I am not sure it is a problem of the `EclipseState` or here.

**Review by joakim-hove — CHANGES_REQUESTED** [review]
This PR uses the raw `Deck` datastructure - outside of opm-common, we must find a way to avoid that. I'm trying to come up with a suggestion.

**Comment by joakim-hove** [comment]
This series of PR's brings in a completely new thing - namely that pinched out cells are activated by the Aquifer code. This is complex code - breaking with an assumption (that cells can only be *de*activated) which has been correct for a long time. This should in my opinion not be merged without proper testing - preferably both unit testing and integration test.

File: opm/grid/cpgrid/processEclipseFormat.cpp
Comment (joakim-hove) [inline opm/grid/cpgrid/processEclipseFormat.cpp:287]: That this line of code actually "turns on" a cell is so special it deserves a comment.

**Comment by joakim-hove** [comment]
> This PR uses the raw Deck datastructure - outside of opm-common, we must find a way to avoid that. I'm trying to come up with a suggestion.

My only suggestion is to have a temporary vector in the numerical aquiferclass with *all* the connections coming from the `AQUCON` keyword, and then a method e.g. `NumericalAquifer::post_process_connections( new_actnum );`

File: opm/grid/cpgrid/processEclipseFormat.cpp
Comment (GitPaean) [reply opm/grid/cpgrid/processEclipseFormat.cpp:287]: Okay. This is not the line turning on a aquifer cell. This line just generates the actnum based on `output`.

Comment (GitPaean) [reply opm/grid/cpgrid/processEclipseFormat.cpp:287]: Here is the place that forcefully keeps aquifer cells active. 

https://github.com/OPM/opm-grid/blob/a28f0bd0dfdf724ee19b3f975efa2c851cd15060/opm/grid/cpgpreprocess/preprocess.c#L354

**Comment by GitPaean** [comment]
> My only suggestion is to have a temporary vector in the numerical aquiferclass with all the connections coming from the AQUCON keyword, and then a method e.g. NumericalAquifer::post_process_connections( new_actnum );

Done, waiting for Jenkins to pass. At the same time, there is one function `processEclipseFormat` using `const Deck&` as input argument already in `CpGridData.hpp`.

**Comment by joakim-hove** [comment]
> At the same time, there is one function processEclipseFormat using const Deck& as input argument already in CpGridData.hpp.

A little surprising - I think it should be removed.

**Review by joakim-hove — APPROVED** [review]

**Pull request merged**
